packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -974,6 +974,23 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
         return stats;
     }
 
+    /**
+     * Gracefully stops the current run of the crawler.
+     *
+     * All the tasks active at the time of calling this method will be allowed to finish.
+     */
+    stop(message: string = 'This crawler has been gracefully stopped.') {
+        // Gracefully starve the this.autoscaledPool, so it doesn't start new tasks. Resolves once the pool is cleared.
+        this.autoscaledPool
+            ?.pause()
+            // Resolves the `autoscaledPool.run()` promise in the `BasicCrawler.run()` method. Since the pool is already paused, it resolves immediately and doesn't kill any tasks.
+            .then(async () => this.autoscaledPool?.abort())
```

Comment:
what's the motivation for not making the method async and ignoring the promise instead? feels handy to be able to await it (and users can always decide themselves to ignore the promise, unlike with the current solution)

also I don't think you need the `async` in here

Reply:
I planned it like this in the beginning, see https://github.com/apify/crawlee/pull/2792/commits/284274bffc1e4153e5afd1447751a4b253925f47

The reasons for moving to this was:
a. parity w/ Python version
b. 
```javascript
async requestHandler({ crawler }) {
    if (iFoundWhatISearchedFor) {
       await requestHandler.stop();
    }
}
```

^ this looks like a completely harmless piece of code, but will cause deadlock. `requestHandler` is waiting for `autoscaledPool.pause`, which is waiting for all tasks to finish (including the aforementioned `requestHandler`).

> also I don't think you need the async in here

You and me both, but [`@typescript-eslint/promise-function-async`](https://typescript-eslint.io/rules/promise-function-async) thinks otherwise.

Reply:
Anyway, I'm definitely for making this `async`, but we need to address these somehow. Also, note that even with the implementation from https://github.com/apify/crawlee/pull/2792/commits/284274bffc1e4153e5afd1447751a4b253925f47
, `crawler.stop` resolves once `autoscaledPool.abort()` has resolved, but that (afaik) doesn't mean the `crawler.run()` has resolved.

Reply:
I'm honestly not sure that there is a good reason for an `async stop()` instead of non-async one.  If you are expected to call it from a request handler, waiting for the crawler to stop there leads to a deadlock and the reasons are quite intuitive.

Should someone need to wait for the crawler to finish outside of the request handler, they can always wait for the promise from `crawler.run()` to resolve.

Reply:
I'm always a bit wary about floating promises, but as long as you guys are fine with it (and there is the catch clause to stop it from killing the whole process), let's see how it goes.

I'll still add the promised tests from #2803 , just to stay on the safer side.